_transcripts/013-rust-1.41.0.md

```diff
@@ -0,0 +1,801 @@
+this change, and I think that RFC is really well written in terms of making
+you understand both why this is necessary and also why this relaxation is
+useful and why it's, sort of, both sufficient and necessary for things that
+people want to do in practice.
+
+And the example that they give in the release notes, I think, is a good one,
+which is imagine that you want to implement, generic over `T`, you want to
+implement `From<YourVec<T>>` for `Vec<T>`.
+
+__Ben__: And `From` is a trait defined in the standard library. As is `Vec`,
+a type defined in the standard library, so you don't own either of those.
+
+__Jon__: Exactly. So this is an instance of the pattern, which is, you
+implement some foreign trait for some foreign type, and previously this was
+just not allowed. You just could not write this. And this is awkward because
+sometimes you really want to say that you can turn my type into some other
+type and you don't control the other type. You don't control the `From` trait,
+and so you don't have any any way to do this. What this RFC proposed that is
+now, sort of, landed on stable is essentially to relax the rules in such a way
+that we can still guarantee that there is only ever one implementation for the
+compiler to choose from. And the way they do that is, to simplify greatly,
+is to say that you are allowed to have an implementation of a foreign trait
+for a foreign type if— you have at least one local type that appears in
+the foreign trait's type arguments and it appears to the left in that list—
+
+__Ben__: it's just the first one of the list, essentially.
+
+__Jon__: It can be multiple, too. Just the leading type parameters of
+the trait have to be local. And if there are any generic parameters that
+appear in the type you're implementing the trait for, then they need to
+be covered. That is, they need to appear in— as type parameters to that
+type. And the rules here might seem somewhat arbitrary, like, why did they
+choose these rules? And the RFC is really good about going through this
+and what "covered" means and why it's necessary. But to try to give some
+intuition here, it doesn't really matter whether your local type is to the
+left or to the right. The RFC just chose one, essentially at random. But the
+reason you want to do this is if every crate is held to the same standard
+of "your local types have to appear to the left," then you can't have two
+crates that accidentally overlap in their implementation. For example,
+imagine that Ben's crate implements `From<HisType<T>>` (??? _needs review_)
+and my crate implements `From<T<MyType>>` then those two implementations
+are technically overlapping because my `T` could be `HisType` or his `T`
+could be `MyType`. And so, as long as the compiler enforces that all the
+local type parameters appear to the left, then these would automatically be
+seen as non-overlapping by the compiler because they couldn't possibly overlap.
```

Comment:
I would like someone to confirm that my interpretation makes sense.

I.e. is the `From<HisType<T>>` and `From<T<MyType>>` faithful to what was meant?

Reply:
Ah, no, I meant `From<(HisType, T)>` and `From<(T, MyType)>`

Reply:
Fixed.